Primitive default values in specifications break the object instantiation

### Describe the bug

Requested situation of Issue https://github.com/fabric8io/kubernetes-client/issues/3869 and fix on Task 
https://github.com/fabric8io/kubernetes-client/pull/3817 lead to a situation where primitive default values in specs lead to generated classes which can't be istantiated.

Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.util.Map (java.lang.String and java.util.Map are in module java.base of loader 'bootstrap')
[INFO]  at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:237)

### Fabric8 Kubernetes Client version

6.2.0

### Steps to reproduce

1.  Change for instance https://github.com/fabric8io/kubernetes-client/blob/master/java-generator/it/src/it/enum-ser-deser/src/test/resources/cert-manager.crds.1.7.1.yaml and add default value to
             spec
              properties:
                description: Desired state of the CertificateRequest resource.
                duration:
                  type: string
                  default: "100"

2. Execute "mvn clean verify -Dinvoker.test=enum-ser-deser"
3. See unmarshall exception during tests.

Generated code for the object flield can not be created, because primitive object is not json.

    @com.fasterxml.jackson.annotation.JsonProperty("duration")
    @com.fasterxml.jackson.annotation.JsonPropertyDescription("The requested 'duration' (i.e. lifetime) of the Certificate. This option may be ignored/overridden by some issuer types.")
    @com.fasterxml.jackson.annotation.JsonSetter(nulls = com.fasterxml.jackson.annotation.Nulls.SKIP)
    private String duration = io.fabric8.kubernetes.client.utils.Serialization.unmarshal("\"100\"", String.class);


Same behaviour for integers or enumerations.

### Expected behavior

Primitive objects may not need io.fabric8.kubernetes.client.utils.Serialization.unmarshal

It would be nice to generate Classes that can be initialized.

### Runtime

other (please specify in additional context)

### Kubernetes API Server version

1.24

### Environment

Windows

### Fabric8 Kubernetes Client Logs

```shell
[ERROR] io.fabric8.it.certmanager.TestEnumSerialization.testDeserialization  Time elapsed: 0.015 s  <<< ERROR!
[INFO] java.lang.IllegalArgumentException:
[INFO] Cannot construct instance of `io.cert_manager.v1.CertificateRequestSpec`, problem: class java.lang.String cannot be cast to class java.util.Map (java.lang.String and java.util.Map are in module java.base of loader 'bootstrap')
[INFO]  at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: io.cert_manager.v1.CertificateRequest["spec"])
[INFO]  at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:4393)
[INFO]  at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:4334)
[INFO]  at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:238)
[INFO]  at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:351)
[INFO]  at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:320)
[INFO]  at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:307)
[INFO]  at io.fabric8.it.certmanager.TestEnumSerialization.testDeserialization(TestEnumSerialization.java:38)
[INFO]  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[INFO]  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[INFO]  at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[INFO]  at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[INFO]  at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
[INFO]  at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
[INFO]  at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
[INFO]  at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
[INFO]  at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
[INFO]  at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
[INFO]  at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
[INFO]  at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
[INFO]  at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
[INFO]  at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
[INFO]  at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
[INFO]  at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
[INFO]  at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
[INFO]  at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
[INFO]  at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)
[INFO]  at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[INFO]  at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
[INFO]  at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
[INFO]  at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
[INFO]  at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
[INFO]  at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
[INFO]  at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
[INFO]  at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
[INFO]  at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
[INFO]  at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
[INFO]  at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
[INFO]  at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
[INFO]  at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
[INFO]  at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
[INFO]  at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
[INFO]  at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
[INFO]  at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
[INFO]  at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
[INFO]  at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
[INFO]  at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
[INFO]  at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
[INFO]  at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
[INFO]  at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
[INFO]  at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
[INFO]  at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
[INFO]  at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
[INFO]  at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
[INFO]  at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
[INFO]  at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
[INFO]  at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
[INFO]  at org.apache.maven.surefire.junitplatform.LazyLauncher.execute(LazyLauncher.java:55)
[INFO]  at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:223)
[INFO]  at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:175)
[INFO]  at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:139)
[INFO]  at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:456)
[INFO]  at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:169)
[INFO]  at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:595)
[INFO]  at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:581)
[INFO] Caused by: com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `io.cert_manager.v1.CertificateRequestSpec`, problem: class java.lang.String cannot be cast to class java.util.Map (java.lang.String and java.util.Map are in module java.base of loader 'bootstrap')
[INFO]  at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: io.cert_manager.v1.CertificateRequest["spec"])
[INFO]  at com.fasterxml.jackson.databind.exc.ValueInstantiationException.from(ValueInstantiationException.java:47)
[INFO]  at com.fasterxml.jackson.databind.DeserializationContext.instantiationException(DeserializationContext.java:2047)
[INFO]  at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.wrapAsJsonMappingException(StdValueInstantiator.java:587)
[INFO]  at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.rewrapCtorProblem(StdValueInstantiator.java:610)
[INFO]  at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createUsingDefault(StdValueInstantiator.java:280)
[INFO]  at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:303)
[INFO]  at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
[INFO]  at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
[INFO]  at io.fabric8.kubernetes.model.jackson.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:134)
[INFO]  at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
[INFO]  at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
[INFO]  at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:4388)
[INFO]  ... 77 more
[INFO] Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.util.Map (java.lang.String and java.util.Map are in module java.base of loader 'bootstrap')
[INFO]  at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:237)
[INFO]  at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:351)
[INFO]  at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:287)
[INFO]  at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:272)
[INFO]  at io.cert_manager.v1.CertificateRequestSpec.<init>(CertificateRequestSpec.java:14)
[INFO]  at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[INFO]  at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
[INFO]  at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[INFO]  at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
[INFO]  at com.fasterxml.jackson.databind.introspect.AnnotatedConstructor.call(AnnotatedConstructor.java:123)
[INFO]  at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createUsingDefault(StdValueInstantiator.java:278)
[INFO]  ... 84 more
```


### Additional context

No additional requirements other then maven and change of one specification file.

@shawkins the change looks promising but I can'z get the master to be build with all Tests. So I can't proof the correctness.

Failures:
[ERROR]   SerializationTest.testSerializeYamlWithAlias:184

Expecting
  Pod(apiVersion=v1, kind=Pod, metadata=ObjectMeta(annotations={}, creationTimestamp=null, deletionGracePeriodSeconds=null, deletionTimestamp=null, finalizers=[], generateName=null, generation=null, labels={}, managedFields=[], name=test-pod-with-alias, namespace=null, ownerReferences=[], resourceVersion=null, selfLink=null, uid=null, additionalProperties={}), spec=PodSpec(activeDeadlineSeconds=null, affinity=null, automountServiceAccountToken=null, containers=[Container(args=[], command=[cat], env=[], envFrom=[], image=ubuntu:bionic, imagePullPolicy=Always, lifecycle=null, livenessProbe=null, name=ubuntu, ports=[], readinessProbe=null, resources=ResourceRequirements(limits={}, requests={cpu=100m}, additionalProperties={}), securityContext=null, startupProbe=null, stdin=null, stdinOnce=null, terminationMessagePath=null, terminationMessagePolicy=null, tty=true, volumeDevices=[], volumeMounts=[], workingDir=null, additionalProperties={}), Container(args=[], command=[cat], env=[], envFrom=[], image=python:3.7, imagePullPolicy=Always, lifecycle=null, livenessProbe=null, name=python3, ports=[], readinessProbe=null, resources=ResourceRequirements(limits={}, requests={cpu=100m}, additionalProperties={}), securityContext=null, startupProbe=null, stdin=null, stdinOnce=null, terminationMessagePath=null, terminationMessagePolicy=null, tty=true, volumeDevices=[], volumeMounts=[], workingDir=null, additionalProperties={})], dnsConfig=null, dnsPolicy=null, enableServiceLinks=null, ephemeralContainers=[], hostAliases=[], hostIPC=null, hostNetwork=null, hostPID=null, hostUsers=null, hostname=null, imagePullSecrets=[], initContainers=[], nodeName=null, nodeSelector={workload=build}, os=null, overhead={}, preemptionPolicy=null, priority=null, priorityClassName=null, readinessGates=[], restartPolicy=null, runtimeClassName=null, schedulerName=null, securityContext=PodSecurityContext(fsGroup=null, fsGroupChangePolicy=null, runAsGroup=1000, runAsNonRoot=null, runAsUser=1000, seLinuxOptions=null, seccompProfile=null, supplementalGroups=[], sysctls=[], windowsOptions=null, additionalProperties={}), serviceAccount=null, serviceAccountName=null, setHostnameAsFQDN=null, shareProcessNamespace=null, subdomain=null, terminationGracePeriodSeconds=null, tolerations=[Toleration(effect=NoSchedule, key=nodeType, operator=Equal, tolerationSeconds=null, value=build, additionalProperties={})], topologySpreadConstraints=[], volumes=[], additionalProperties={}), status=null, additionalProperties={})
to have a property or a field named "spec.tolerations.size"
(static and synthetic fields are ignored)
        at io.fabric8.kubernetes.client.utils.SerializationTest.testSerializeYamlWithAlias(SerializationTest.java:184)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@
 
 #### Improvements
 * Fix #4355: for exec, attach, upload, and copy operations the container id/name will be validated or chosen prior to the remote call.  You may also use the kubectl.kubernetes.io/default-container annotation to specify the default container.
+*  Fix #4530: generalizing the Serialization logic to allow for primitive values and clarifying the type expectations.
 
 #### Dependency Upgrade
 

--- a/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java
+++ b/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java
@@ -286,10 +286,10 @@ public interface KubernetesClient extends Client {
   NonNamespaceOperation<ComponentStatus, ComponentStatusList, Resource<ComponentStatus>> componentstatuses();
 
   /**
-   * Load a Kubernetes resource object from file InputStream
+   * Load Kubernetes resource object(s) from the provided InputStream.
    *
-   * @param is File input stream object containing json/yaml content
-   * @return deserialized object
+   * @param is the input stream containing JSON/YAML content
+   * @return an operation instance to work on the list of Kubernetes Resource objects
    */
   ParameterNamespaceListVisitFromServerGetDeleteRecreateWaitApplicable<HasMetadata> load(InputStream is);
 
@@ -297,7 +297,7 @@ public interface KubernetesClient extends Client {
    * Load a Kubernetes list object
    *
    * @param s kubernetes list as string
-   * @return deserialized KubernetesList object
+   * @return an operation instance to work on the deserialized KubernetesList objects
    */
   ParameterNamespaceListVisitFromServerGetDeleteRecreateWaitApplicable<HasMetadata> resourceList(String s);
 
@@ -340,10 +340,19 @@ public interface KubernetesClient extends Client {
    * KubernetesResource operations. You can pass any Kubernetes resource as string object and do
    * all operations
    *
-   * @param s Kubernetes resource object as string
+   * @param s a Kubernetes resource object as string
    * @return operations object for Kubernetes resource
    */
   NamespaceableResource<HasMetadata> resource(String s);
+
+  /**
+   * KubernetesResource operations. You can pass any Kubernetes resource as an InputStream object and perform
+   * all operations
+   *
+   * @param is the InputStream containing a serialized Kubernetes resource.
+   * @return operations object for Kubernetes resource.
+   */
+  NamespaceableResource<HasMetadata> resource(InputStream is);
 
   /**
    * Operations for Binding resource in APIgroup core/v1
@@ -514,7 +523,7 @@ public interface KubernetesClient extends Client {
 
   /**
    * Visit all resources with the given {@link ApiVisitor}.
-   * 
+   *
    * @param visitor
    */
   void visitResources(ApiVisitor visitor);

--- a/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/NamespacedKubernetesClientAdapter.java
+++ b/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/NamespacedKubernetesClientAdapter.java
@@ -284,6 +284,11 @@ public class NamespacedKubernetesClientAdapter<N extends NamespacedKubernetesCli
   }
 
   @Override
+  public NamespaceableResource<HasMetadata> resource(InputStream is) {
+    return getClient().resource(is);
+  }
+
+  @Override
   public MixedOperation<Binding, KubernetesResourceList<Binding>, Resource<Binding>> bindings() {
     return getClient().bindings();
   }

--- a/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java
+++ b/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/Serialization.java
@@ -23,6 +23,7 @@ import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
 import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator;
 import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
 import io.fabric8.kubernetes.api.model.KubernetesResource;
+import io.fabric8.kubernetes.api.model.runtime.RawExtension;
 import io.fabric8.kubernetes.client.KubernetesClientException;
 import io.fabric8.kubernetes.model.jackson.UnmatchedFieldTypeModule;
 import org.yaml.snakeyaml.DumperOptions;
@@ -158,6 +159,8 @@ public class Serialization {
 
   /**
    * Unmarshals a stream.
+   * <p>
+   * The type is assumed to be {@link KubernetesResource}
    *
    * @param is The {@link InputStream}.
    * @param <T> The target type.
@@ -170,6 +173,8 @@ public class Serialization {
 
   /**
    * Unmarshals a stream optionally performing placeholder substitution to the stream.
+   * <p>
+   * The type is assumed to be {@link KubernetesResource}
    *
    * @param is The {@link InputStream}.
    * @param parameters A {@link Map} with parameters for placeholder substitution.
@@ -190,6 +195,8 @@ public class Serialization {
 
   /**
    * Unmarshals a stream.
+   * <p>
+   * The type is assumed to be {@link KubernetesResource}
    *
    * @param is The {@link InputStream}.
    * @param mapper The {@link ObjectMapper} to use.
@@ -202,6 +209,8 @@ public class Serialization {
 
   /**
    * Unmarshals a stream optionally performing placeholder substitution to the stream.
+   * <p>
+   * The type is assumed to be {@link KubernetesResource}
    *
    * @param is The {@link InputStream}.
    * @param mapper The {@link ObjectMapper} to use.
@@ -231,11 +240,15 @@ public class Serialization {
       bis.reset();
 
       final T result;
-      if (intch != '{') {
+      if (intch != '{' && intch != '[') {
         final Yaml yaml = new Yaml(new SafeConstructor(), new Representer(), new DumperOptions(),
             new CustomYamlTagResolverWithLimit());
-        final Map<String, Object> obj = yaml.load(bis);
-        result = mapper.convertValue(obj, type);
+        final Object obj = yaml.load(bis);
+        if (obj instanceof Map) {
+          result = mapper.convertValue(obj, type);
+        } else {
+          result = mapper.convertValue(new RawExtension(obj), type);
+        }
       } else {
         result = mapper.readerFor(type).readValue(bis);
       }
@@ -247,6 +260,8 @@ public class Serialization {
 
   /**
    * Unmarshals a {@link String}
+   * <p>
+   * The type is assumed to be {@link KubernetesResource}
    *
    * @param str The {@link String}.
    * @param <T> template argument denoting type

--- a/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/utils/SerializationSingleDocumentUnmarshalTest.java
+++ b/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/utils/SerializationSingleDocumentUnmarshalTest.java
@@ -27,24 +27,28 @@ import static org.assertj.core.api.Assertions.assertThat;
 
 class SerializationSingleDocumentUnmarshalTest {
   @ParameterizedTest(name = "#{index} - unmarshal {0}")
-  @ValueSource(strings = {"document-with-trailing-document-delimiter.yml", "document-with-leading-document-delimiter.yml", "document-with-leading-and-trailing-document-delimiter.yml", "document-with-no-document-delimiter.yml"})
+  @ValueSource(strings = {
+      "document-with-trailing-document-delimiter.yml",
+      "document-with-leading-document-delimiter.yml",
+      "document-with-leading-and-trailing-document-delimiter.yml",
+      "document-with-no-document-delimiter.yml"
+  })
   void unmarshalWithSingleDocumentWithDocumentDelimiterShouldReturnKubernetesResource(String arg) {
     // When
     final KubernetesResource result = Serialization.unmarshal(
-      SerializationTest.class.getResourceAsStream(String.format("/serialization/%s", arg)),
-      Collections.emptyMap()
-    );
+        SerializationTest.class.getResourceAsStream(String.format("/serialization/%s", arg)),
+        Collections.emptyMap());
     // Then
     assertThat(result)
-      .asInstanceOf(InstanceOfAssertFactories.type(Service.class))
-      .hasFieldOrPropertyWithValue("apiVersion", "v1")
-      .hasFieldOrPropertyWithValue("kind", "Service")
-      .hasFieldOrPropertyWithValue("metadata.name", "redis-master")
-      .hasFieldOrPropertyWithValue("metadata.labels.app", "redis")
-      .hasFieldOrPropertyWithValue("metadata.labels.tier", "backend")
-      .hasFieldOrPropertyWithValue("metadata.labels.role", "master")
-      .hasFieldOrPropertyWithValue("spec.selector.app", "redis")
-      .hasFieldOrPropertyWithValue("spec.selector.tier", "backend")
-      .hasFieldOrPropertyWithValue("spec.selector.role", "master");
+        .asInstanceOf(InstanceOfAssertFactories.type(Service.class))
+        .hasFieldOrPropertyWithValue("apiVersion", "v1")
+        .hasFieldOrPropertyWithValue("kind", "Service")
+        .hasFieldOrPropertyWithValue("metadata.name", "redis-master")
+        .hasFieldOrPropertyWithValue("metadata.labels.app", "redis")
+        .hasFieldOrPropertyWithValue("metadata.labels.tier", "backend")
+        .hasFieldOrPropertyWithValue("metadata.labels.role", "master")
+        .hasFieldOrPropertyWithValue("spec.selector.app", "redis")
+        .hasFieldOrPropertyWithValue("spec.selector.tier", "backend")
+        .hasFieldOrPropertyWithValue("spec.selector.role", "master");
   }
 }

--- a/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/utils/SerializationTest.java
+++ b/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/utils/SerializationTest.java
@@ -52,6 +52,9 @@ import org.assertj.core.api.InstanceOfAssertFactories;
 import org.assertj.core.groups.Tuple;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
 
 import java.io.File;
 import java.io.IOException;
@@ -65,10 +68,12 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
 import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertThrows;
 
 class SerializationTest {
 
@@ -195,7 +200,7 @@ class SerializationTest {
   }
 
   @Test
-  void testClone() {
+  void cloneKubernetesResourceReturnsDifferentInstance() {
     // Given
     Pod pod = new PodBuilder().withNewMetadata().withName("pod").endMetadata().build();
     // When
@@ -208,7 +213,7 @@ class SerializationTest {
   }
 
   @Test
-  void testCloneNonResource() {
+  void cloneNonResourceReturnsDifferentInstance() {
     // Given
     Map<String, String> value = Collections.singletonMap("key", "value");
     // When
@@ -237,7 +242,7 @@ class SerializationTest {
   }
 
   @Test
-  void unmarshalWithInvalidYamlShouldThrowException() {
+  void unmarshalWithInvalidYamlShouldReturnRawExtension() {
     // Given
     final InputStream is = SerializationTest.class.getResourceAsStream("/serialization/invalid-yaml.yml");
     // When
@@ -247,27 +252,59 @@ class SerializationTest {
   }
 
   @Test
-  void unmarshalArrays() {
+  void unmarshalYamlArrayShouldThrowException() {
+    assertThatIllegalArgumentException()
+        .isThrownBy(() -> Serialization.unmarshal("- 1\n- 2"))
+        .withMessageStartingWith("Cannot parse a nested array containing non-object resource");
+  }
+
+  @Test
+  void unmarshalJsonArrayShouldThrowException() {
+    assertThatExceptionOfType(KubernetesClientException.class)
+        .isThrownBy(() -> Serialization.unmarshal("[1, 2]"))
+        .withMessage("An error has occurred.")
+        .havingCause()
+        .withMessageStartingWith("Cannot parse a nested array containing non-object resource");
+  }
+
+  @Test
+  void unmarshalYamlArrayWithProvidedTypeShouldDeserialize() {
     // not valid as KubernetesResource - we'd have to look ahead to know if the array values
     // were not hasmetadata
-    assertThrows(KubernetesClientException.class, () -> Serialization.unmarshal("[1, 2]"));
-    assertThrows(IllegalArgumentException.class, () -> Serialization.unmarshal("- 1\n- 2"));
-
-    assertEquals(Arrays.asList(1, 2), Serialization.unmarshal("[1, 2]", List.class));
     assertEquals(Arrays.asList(1, 2), Serialization.unmarshal("- 1\n- 2", List.class));
   }
 
   @Test
-  void unmarshalPrimitives() {
-    // as json
-    RawExtension raw = Serialization.unmarshal("\"a\"");
-    assertEquals("a", raw.getValue());
-    // as yaml
-    raw = Serialization.unmarshal("a");
-    assertEquals("a", raw.getValue());
+  void unmarshalJsonArrayWithProvidedTypeShouldDeserialize() {
+    // not valid as KubernetesResource - we'd have to look ahead to know if the array values
+    // were not hasmetadata
+    assertEquals(Arrays.asList(1, 2), Serialization.unmarshal("[1, 2]", List.class));
+  }
 
+  @ParameterizedTest(name = "''{0}'' should be deserialized as ''{1}''")
+  @MethodSource("unmarshalPrimitivesInput")
+  void unmarshalPrimitives(String input, Object expected) {
+    assertThat(Serialization.<RawExtension> unmarshal(input))
+        .extracting(RawExtension::getValue)
+        .isEqualTo(expected);
     assertEquals("a", Serialization.unmarshal("\"a\"", String.class));
     assertEquals("a", Serialization.unmarshal("a", String.class));
+  }
+
+  @ParameterizedTest(name = "''{0}'' and ''{2}'' target type should be deserialized as ''{1}''")
+  @MethodSource("unmarshalPrimitivesInput")
+  void unmarshalPrimitivesWithType(String input, Object expected, Class<?> targetType) {
+    assertThat(Serialization.unmarshal(input, targetType))
+        .isEqualTo(expected);
+  }
+
+  static Stream<Arguments> unmarshalPrimitivesInput() {
+    return Stream.of(
+        Arguments.arguments("\"a\"", "a", String.class), // JSON
+        Arguments.arguments("a", "a", String.class), // YAML
+        Arguments.arguments("1", 1, Integer.class),
+        Arguments.arguments("true", true, Boolean.class),
+        Arguments.arguments("1.2", 1.2, Double.class));
   }
 
   @Test

--- a/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/utils/SerializationTest.java
+++ b/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/utils/SerializationTest.java
@@ -45,6 +45,7 @@ import io.fabric8.kubernetes.api.model.coordination.v1.Lease;
 import io.fabric8.kubernetes.api.model.coordination.v1.LeaseSpec;
 import io.fabric8.kubernetes.api.model.runtime.RawExtension;
 import io.fabric8.kubernetes.client.CustomResource;
+import io.fabric8.kubernetes.client.KubernetesClientException;
 import io.fabric8.kubernetes.model.annotation.Group;
 import io.fabric8.kubernetes.model.annotation.Version;
 import org.assertj.core.api.InstanceOfAssertFactories;
@@ -61,6 +62,7 @@ import java.time.ZonedDateTime;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
 
@@ -235,15 +237,37 @@ class SerializationTest {
   }
 
   @Test
-  @DisplayName("unmarshal, with invalid YAML content, should throw Exception")
   void unmarshalWithInvalidYamlShouldThrowException() {
     // Given
     final InputStream is = SerializationTest.class.getResourceAsStream("/serialization/invalid-yaml.yml");
     // When
-    final ClassCastException result = assertThrows(ClassCastException.class, () -> Serialization.unmarshal(is));
+    RawExtension result = Serialization.unmarshal(is);
     // Then
-    assertThat(result)
-        .hasMessageContainingAll("java.lang.String", "cannot be cast to", "java.util.Map");
+    assertEquals("This\nis not\nYAML", result.getValue());
+  }
+
+  @Test
+  void unmarshalArrays() {
+    // not valid as KubernetesResource - we'd have to look ahead to know if the array values
+    // were not hasmetadata
+    assertThrows(KubernetesClientException.class, () -> Serialization.unmarshal("[1, 2]"));
+    assertThrows(IllegalArgumentException.class, () -> Serialization.unmarshal("- 1\n- 2"));
+
+    assertEquals(Arrays.asList(1, 2), Serialization.unmarshal("[1, 2]", List.class));
+    assertEquals(Arrays.asList(1, 2), Serialization.unmarshal("- 1\n- 2", List.class));
+  }
+
+  @Test
+  void unmarshalPrimitives() {
+    // as json
+    RawExtension raw = Serialization.unmarshal("\"a\"");
+    assertEquals("a", raw.getValue());
+    // as yaml
+    raw = Serialization.unmarshal("a");
+    assertEquals("a", raw.getValue());
+
+    assertEquals("a", Serialization.unmarshal("\"a\"", String.class));
+    assertEquals("a", Serialization.unmarshal("a", String.class));
   }
 
   @Test

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/impl/KubernetesClientImpl.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/impl/KubernetesClientImpl.java
@@ -357,12 +357,28 @@ public class KubernetesClientImpl extends BaseClient implements NamespacedKubern
     return new NamespaceableResourceAdapter<>(item, op);
   }
 
+  private NamespaceableResource<HasMetadata> resource(Object resource) {
+    if (resource instanceof HasMetadata) {
+      return resource((HasMetadata) resource);
+    }
+    throw new KubernetesClientException("Unable to create a valid resource from the provided object (" +
+        resource.getClass().getName() + ")");
+  }
+
   /**
    * {@inheritDoc}
    */
   @Override
   public NamespaceableResource<HasMetadata> resource(String s) {
-    return resource((HasMetadata) Serialization.unmarshal(s));
+    return resource(Serialization.<Object> unmarshal(s));
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public NamespaceableResource<HasMetadata> resource(InputStream is) {
+    return resource(Serialization.<Object> unmarshal(is));
   }
 
   /**

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/impl/KubernetesClientImplTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/impl/KubernetesClientImplTest.java
@@ -16,11 +16,14 @@
 package io.fabric8.kubernetes.client.impl;
 
 import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.Pod;
 import io.fabric8.kubernetes.api.model.Service;
 import io.fabric8.kubernetes.api.model.apps.Deployment;
 import io.fabric8.kubernetes.client.Config;
 import io.fabric8.kubernetes.client.ConfigBuilder;
 import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.dsl.ParameterNamespaceListVisitFromServerGetDeleteRecreateWaitApplicable;
 import io.fabric8.kubernetes.client.http.BasicBuilder;
 import io.fabric8.kubernetes.client.http.HttpHeaders;
 import io.fabric8.kubernetes.client.utils.HttpClientUtils;
@@ -35,6 +38,7 @@ import org.mockito.Mockito;
 import java.io.ByteArrayInputStream;
 import java.io.File;
 import java.io.FileInputStream;
+import java.io.IOException;
 import java.io.InputStream;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
@@ -44,6 +48,8 @@ import java.util.List;
 import java.util.Map;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
 import static org.junit.jupiter.api.Assertions.assertArrayEquals;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.fail;
@@ -152,5 +158,66 @@ class KubernetesClientImplTest {
     assertEquals("a", currentConfig.getImpersonateUsername());
     assertArrayEquals(new String[] { "b" }, currentConfig.getImpersonateGroups());
     assertEquals(Collections.singletonList("d"), currentConfig.getImpersonateExtras().get("c"));
+  }
+
+  @Test
+  @DisplayName("resource(String).get with HasMetadata should deserialize")
+  void resourceFromStringWithHasMetadata() {
+    assertThat(new KubernetesClientImpl().resource("apiVersion: v1\nkind: Pod").get())
+        .isInstanceOf(Pod.class);
+  }
+
+  @Test
+  @DisplayName("resource(String) with no HasMetadata should throwException")
+  void resourceFromStringWithInvalid() {
+    final KubernetesClient kc = new KubernetesClientImpl();
+    assertThatExceptionOfType(KubernetesClientException.class)
+        .isThrownBy(() -> kc.resource("NotAPod"))
+        .withMessageStartingWith("Unable to create a valid resource from the provided object");
+  }
+
+  @Test
+  @DisplayName("resource(InputStream).get with HasMetadata should deserialize")
+  void resourceFromInputStreamWithHasMetadata() throws IOException {
+    final String podYaml = "apiVersion: v1\nkind: Pod";
+    try (InputStream is = new ByteArrayInputStream(podYaml.getBytes(StandardCharsets.UTF_8))) {
+      assertThat(new KubernetesClientImpl().resource(is).get())
+          .isInstanceOf(Pod.class);
+    }
+  }
+
+  @Test
+  @DisplayName("resource(InputStream) with no HasMetadata should throwException")
+  void resourceFromInputStreamWithInvalid() throws IOException {
+    final KubernetesClient kc = new KubernetesClientImpl();
+    final String podYaml = "NotAPod";
+    try (InputStream is = new ByteArrayInputStream(podYaml.getBytes(StandardCharsets.UTF_8))) {
+      assertThatExceptionOfType(KubernetesClientException.class)
+          .isThrownBy(() -> kc.resource(is))
+          .withMessageStartingWith("Unable to create a valid resource from the provided object");
+    }
+  }
+
+  @Test
+  @DisplayName("load(InputStream).get with HasMetadata should deserialize")
+  void loadFromInputStreamWithHasMetadata() throws IOException {
+    final String podYaml = "apiVersion: v1\nkind: Pod";
+    try (InputStream is = new ByteArrayInputStream(podYaml.getBytes(StandardCharsets.UTF_8))) {
+      assertThat(new KubernetesClientImpl().load(is).get())
+          .containsExactly(new Pod());
+    }
+  }
+
+  @Test
+  @DisplayName("load(InputStream).get with no HasMetadata should throwException")
+  void loadFromInputStreamWithInvalid() throws IOException {
+    final String podYaml = "NotAPod";
+    try (InputStream is = new ByteArrayInputStream(podYaml.getBytes(StandardCharsets.UTF_8))) {
+      final ParameterNamespaceListVisitFromServerGetDeleteRecreateWaitApplicable<HasMetadata> load = new KubernetesClientImpl()
+          .load(is);
+      assertThatIllegalArgumentException()
+          .isThrownBy(load::get)
+          .withMessageStartingWith("Could not convert item to a list of HasMetadata");
+    }
   }
 }

--- a/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/runtime/RawExtension.java
+++ b/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/runtime/RawExtension.java
@@ -21,7 +21,9 @@ import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
 import io.fabric8.kubernetes.api.model.AnyType;
 import io.fabric8.kubernetes.api.model.KubernetesResource;
 import io.sundr.builder.annotations.Buildable;
+import lombok.ToString;
 
+@ToString(callSuper = true)
 @JsonDeserialize(using = com.fasterxml.jackson.databind.JsonDeserializer.None.class)
 @Buildable(editableEnabled = false, validationEnabled = false, generateBuilderPackage = true, lazyCollectionInitEnabled = false, builderPackage = "io.fabric8.kubernetes.api.builder")
 public class RawExtension extends AnyType implements KubernetesResource {

--- a/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java
+++ b/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java
@@ -89,9 +89,12 @@ public class KubernetesDeserializer extends JsonDeserializer<KubernetesResource>
       return fromObjectNode(jp, node);
     } else if (node.isArray()) {
       return fromArrayNode(jp, node);
-    } else {
+    }
+    Object object = node.traverse(jp.getCodec()).readValueAs(Object.class);
+    if (object == null) {
       return null;
     }
+    return new RawExtension(object);
   }
 
   private KubernetesResource fromArrayNode(JsonParser jp, JsonNode node) throws IOException {
@@ -105,6 +108,8 @@ public class KubernetesDeserializer extends JsonDeserializer<KubernetesResource>
           throw new JsonMappingException(jp, "Cannot parse a nested array containing a non-HasMetadata resource");
         }
         list.add((HasMetadata) resource);
+      } else {
+        throw new JsonMappingException(jp, "Cannot parse a nested array containing non-object resource");
       }
     }
     return new KubernetesListBuilder().withItems(list).build();

